File: backend/src/routes/jobs.js
/**
 * Job Description Routes — Save JD, Semantic Match, Cover Letter
 */

import { Router } from 'express';
import prisma from '../database.js';
import { authenticate } from '../middleware/auth.js';
import { semanticMatchScore } from '../services/semantic/matcher.js';
import { generateCoverLetter } from '../services/ai/coverLetterGen.js';

const router = Router();

/**
 * POST /api/job-descriptions — Save a job description
 */
router.post('/job-descriptions', authenticate, async (req, res, next) => {
  try {
    const { title, company, rawText } = req.body;

    if (!rawText) {
      return res.status(400).json({ error: 'Job description text is required' });
    }

    const jd = await prisma.jobDescription.create({
      data: {
        userId: req.user.id,
        title: title || 'Untitled Position',
        company: company || null,
        rawText,
      },
    });

    res.status(201).json({
      id: jd.id,
      title: jd.title,
      company: jd.company,
      createdAt: jd.createdAt,
    });
  } catch (err) {
    next(err);
  }
});

/**
 * GET /api/job-descriptions — List user's saved JDs
 */
router.get('/job-descriptions', authenticate, async (req, res, next) => {
  try {
    const jds = await prisma.jobDescription.findMany({
      where: { userId: req.user.id },
      orderBy: { createdAt: 'desc' },
      select: { id: true, title: true, company: true, createdAt: true },
    });
    res.json(jds);
  } catch (err) {
    next(err);
  }
});

/**
 * DELETE /api/job-descriptions/:id — Delete a saved job description
 */
router.delete('/job-descriptions/:id', authenticate, async (req, res, next) => {
  try {
    const jd = await prisma.jobDescription.findFirst({
      where: { id: req.params.id, userId: req.user.id },
    });

    if (!jd) {
      return res.status(404).json({ error: 'Job description not found' });
    }

    await prisma.jobDescription.delete({
      where: { id: req.params.id },
    });

    res.json({ message: 'Job description deleted successfully', id: req.params.id });
  } catch (err) {
    next(err);
  }
});

/**
 * POST /api/resumes/:id/match/:jdId — Semantic match score + missing keywords
 */
router.post('/resumes/:id/match/:jdId', authenticate, async (req, res, next) => {
  try {
    const resume = await prisma.resume.findFirst({
      where: { id: req.params.id, userId: req.user.id },
    });
    const jd = await prisma.jobDescription.findFirst({
      where: { id: req.params.jdId, userId: req.user.id },
    });

    if (!resume || !jd) {
      return res.status(404).json({ error: 'Resume or job description not found' });
    }

    const result = await semanticMatchScore(resume.rawText, jd.rawText);
    res.json(result);
  } catch (err) {
    next(err);
  }
});

/**
 * POST /api/resumes/:id/cover-letter/:jdId — Generate cover letter
 */
router.post('/resumes/:id/cover-letter/:jdId', authenticate, async (req, res, next) => {
  try {
    const resume = await prisma.resume.findFirst({
      where: { id: req.params.id, userId: req.user.id },
    });
    const jd = await prisma.jobDescription.findFirst({
      where: { id: req.params.jdId, userId: req.user.id },
    });

    if (!resume || !jd) {
      return res.status(404).json({ error: 'Resume or job description not found' });
    }

    const coverLetter = await generateCoverLetter(resume.rawText, jd.rawText, resume.parsedJson);
    res.json(coverLetter);
  } catch (err) {
    next(err);
  }
});

export default router;

File: backend/app/routes/jobs.py
"""
ResumeIQ — Job Description, Semantic Matching & Cover Letter Routes (FastAPI)
"""

from fastapi import APIRouter, Depends, HTTPException, status
from pydantic import BaseModel
from sqlalchemy.orm import Session
from typing import Optional, List

from ..database import get_db, User, JobDescription, Resume
from ..middleware.auth import get_current_user
from ..services.ml.semantic_matcher import semantic_match_score
from ..services.ml.cover_letter_gen import generate_cover_letter

router = APIRouter(prefix="", tags=["Job Descriptions"])

class CreateJDRequest(BaseModel):
    title: Optional[str] = "Untitled Position"
    company: Optional[str] = None
    rawText: str

@router.post("/job-descriptions", status_code=status.HTTP_201_CREATED)
@router.post("/v1/job-descriptions", status_code=status.HTTP_201_CREATED)
def create_job_description(
    payload: CreateJDRequest,
    user: User = Depends(get_current_user),
    db: Session = Depends(get_db)
):
    if not payload.rawText or not payload.rawText.strip():
        raise HTTPException(
            status_code=status.HTTP_400_BAD_REQUEST,
            detail="Job description text is required."
        )

    jd = JobDescription(
        userId=user.id,
        title=payload.title or "Untitled Position",
        company=payload.company,
        rawText=payload.rawText.strip(),
    )
    db.add(jd)
    db.commit()
    db.refresh(jd)

    return {
        "id": jd.id,
        "title": jd.title,
        "company": jd.company,
        "createdAt": jd.createdAt.isoformat() if jd.createdAt else None,
    }

@router.get("/job-descriptions")
@router.get("/v1/job-descriptions")
def list_job_descriptions(
    user: User = Depends(get_current_user),
    db: Session = Depends(get_db)
):
    jds = (
        db.query(JobDescription)
        .filter(JobDescription.userId == user.id)
        .order_by(JobDescription.createdAt.desc())
        .all()
    )

    return [
        {
            "id": j.id,
            "title": j.title,
            "company": j.company,
            "createdAt": j.createdAt.isoformat() if j.createdAt else None,
        }
        for j in jds
    ]

@router.delete("/job-descriptions/{jd_id}")
@router.delete("/v1/job-descriptions/{jd_id}")
def delete_job_description(
    jd_id: str,
    user: User = Depends(get_current_user),
    db: Session = Depends(get_db)
):
    jd = db.query(JobDescription).filter(JobDescription.id == jd_id, JobDescription.userId == user.id).first()
    if not jd:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail="Job description not found.")

    db.delete(jd)
    db.commit()

    return {"message": "Job description deleted successfully", "id": jd_id}

@router.post("/resumes/{resume_id}/match/{jd_id}")
@router.post("/v1/resumes/{resume_id}/match/{jd_id}")
def match_resume_to_jd(
    resume_id: str,
    jd_id: str,
    user: User = Depends(get_current_user),
    db: Session = Depends(get_db)
):
    resume = db.query(Resume).filter(Resume.id == resume_id, Resume.userId == user.id).first()
    jd = db.query(JobDescription).filter(JobDescription.id == jd_id, JobDescription.userId == user.id).first()

    if not resume or not jd:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="Resume or job description not found."
        )

    return semantic_match_score(resume.rawText, jd.rawText)

@router.post("/resumes/{resume_id}/cover-letter/{jd_id}")
@router.post("/v1/resumes/{resume_id}/cover-letter/{jd_id}")
def create_cover_letter(
    resume_id: str,
    jd_id: str,
    user: User = Depends(get_current_user),
    db: Session = Depends(get_db)
):
    resume = db.query(Resume).filter(Resume.id == resume_id, Resume.userId == user.id).first()
    jd = db.query(JobDescription).filter(JobDescription.id == jd_id, JobDescription.userId == user.id).first()

    if not resume or not jd:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="Resume or job description not found."
        )

    return generate_cover_letter(resume.rawText, jd.rawText, resume.parsedJson)
